src/System.Linq/tests/SelectTests.cs

```diff
@@ -0,0 +1,590 @@
+            {
+                var expected = selector(selector(source[index]));
+                Assert.Equal(expected, item);
+                index++;
+            }
+
+            Assert.Equal(source.Length, index);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsAList_ReturnsExpectedValues()
+        {
+            List<int> source = new List<int> { 1, 2, 3, 4, 5 };
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                var expected = selector(selector(source[index]));
+                Assert.Equal(expected, item);
+                index++;
+            }
+
+            Assert.Equal(source.Count, index);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsIReadOnlyCollection_ReturnsExpectedValues()
+        {
+            IReadOnlyCollection<int> source = new ReadOnlyCollection<int>(new List<int> { 1, 2, 3, 4, 5 });
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(selector(index));
+                Assert.Equal(expected, item);
+            }
+
+            Assert.Equal(source.Count, index);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsICollection_ReturnsExpectedValues()
+        {
+            ICollection<int> source = new LinkedList<int>(new List<int> { 1, 2, 3, 4, 5 });
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(selector(index));
+                Assert.Equal(expected, item);
+            }
+
+            Assert.Equal(source.Count, index);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsIEnumerable_ReturnsExpectedValues()
+        {
+            int nbOfItems = 5;
+            IEnumerable<int> source = Enumerable.Range(1, 5);
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(selector(index));
+                Assert.Equal(expected, item);
+            }
+
+            Assert.Equal(nbOfItems, index);
+        }
+
+        [Fact]
+        public void Select_SourceIsEmptyEnumerable_ReturnedCollectionHasNoElements()
+        {
+            IEnumerable<int> source = Enumerable.Empty<int>();
+
+            IEnumerable<int> result = source.Select(i => i + 1);
```

Comment:
Consider also setting a captured Boolean flag in the delegate and then asserting below that it's not getting set, in order to verify not only that Select didn't yield any results but also never executed the delegate.